core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -1,18 +1,1604 @@
+
+  /**
+   * Evaluate each effect in the structure in parallel, and collect
+   * the results. For a sequential version, see `collectAll`.
+   */
+  final def collectAllPar[R >: LowerR, E <: UpperE, A](as: Iterable[ZIO[R, E, A]]): ZIO[R, E, List[A]] =
+    foreachPar[R, E, ZIO[R, E, A], A](as)(identity(_))
+
+  /**
+   * Evaluate each effect in the structure in parallel, and collect
+   * the results. For a sequential version, see `collectAll`.
+   *
+   * Unlike `foreachAllPar`, this method will use at most `n` fibers.
+   */
+  final def collectAllParN[R >: LowerR, E <: UpperE, A](n: Long)(as: Iterable[ZIO[R, E, A]]): ZIO[R, E, List[A]] =
+    foreachParN[R, E, ZIO[R, E, A], A](n)(as)(identity(_))
+
+  /**
+   * Races an `IO[E, A]` against zero or more other effects. Yields either the
+   * first success or the last failure.
+   */
+  final def raceAll[R >: LowerR, R1 >: LowerR <: R, E <: UpperE, A](
+    zio: ZIO[R, E, A],
+    ios: Iterable[ZIO[R1, E, A]]
+  ): ZIO[R1, E, A] =
+    ios.foldLeft[ZIO[R1, E, A]](zio)(_ race _)
+
+  /**
+   * Reduces an `Iterable[IO]` to a single `IO`, working in parallel.
+   */
+  final def reduceAll[R >: LowerR, R1 >: LowerR <: R, E <: UpperE, A](a: ZIO[R, E, A], as: Iterable[ZIO[R1, E, A]])(
+    f: (A, A) => A
+  ): ZIO[R1, E, A] =
+    as.foldLeft[ZIO[R1, E, A]](a) { (l, r) =>
+      l.zipPar(r).map(f.tupled)
+    }
+
+  /**
+   * Merges an `Iterable[IO]` to a single IO, working in parallel.
+   */
+  final def mergeAll[R >: LowerR, E <: UpperE, A, B](
+    in: Iterable[ZIO[R, E, A]]
+  )(zero: B)(f: (B, A) => B): ZIO[R, E, B] =
+    in.foldLeft[ZIO[R, E, B]](succeedLazy[B](zero))((acc, a) => acc.zipPar(a).map(f.tupled))
+
+  /**
+   * Strictly-evaluated unit lifted into the `ZIO` monad.
+   */
+  final def unit[R >: LowerR]: ZIO[R, Nothing, Unit] = succeed(())
+
+  /**
+   * The moral equivalent of `if (p) exp`
+   */
+  final def when[R >: LowerR, E <: UpperE](b: Boolean)(zio: ZIO[R, E, Unit]): ZIO[R, E, Unit] =
+    if (b) zio else unit
+
+  /**
+   * The moral equivalent of `if (p) exp` when `p` has side-effects
+   */
+  final def whenM[R >: LowerR, E <: UpperE](b: ZIO[R, E, Boolean])(zio: ZIO[R, E, Unit]): ZIO[R, E, Unit] =
+    b.flatMap(b => if (b) zio else unit)
+
+  /**
+   * Folds an `Iterable[A]` using an effectful function `f`, working sequentially.
+   */
+  final def foldLeft[E, S, A](in: Iterable[A])(zero: S)(f: (S, A) => IO[E, S]): IO[E, S] =
+    in.foldLeft(IO.succeed(zero): IO[E, S]) { (acc, el) =>
+      acc.flatMap(f(_, el))
+    }
+
+  /**
+   * Returns information about the current fiber, such as its fiber identity.
+   */
+  final def descriptor: UIO[Fiber.Descriptor] = ZIO.Descriptor
 }
-object ZIO extends Serializable {
-  final def apply[F[_, _]](implicit F: ZIO[F]): ZIO[F] = F
 
-  implicit final val ZIOIO: ZIO[IO] = new ZIO[IO] {
-    final def liftZIO[E, A](io: IO[E, A]): IO[E, A] = io
+trait ZIO_E_Any extends ZIO_E_Throwable {
```

Comment:
I am not a fan of these strange trait names...

Reply:
I'll clean them up in followup PR.